include/flux/op/fill.hpp

```diff
@@ -6,18 +6,50 @@
 #define FLUX_OP_FILL_HPP_INCLUDED
 
 #include <flux/op/for_each.hpp>
+#include <type_traits>
+#include <cstring>
 
 namespace flux {
 
 namespace detail {
 
 struct fill_fn {
+private:
+    template <typename Seq, typename Value>
+    static constexpr auto impl(Seq& seq, Value const& value)
+    {
+        flux::for_each(seq, [&value](auto&& elem) { FLUX_FWD(elem) = value; });
+    }
+
+public:
     template <typename Value, writable_sequence_of<Value> Seq>
     constexpr void operator()(Seq&& seq, Value const& value) const
     {
-        flux::for_each(seq, [&value](auto&& elem) {
-            FLUX_FWD(elem) = value;
-        });
+        constexpr bool can_memset = 
+            contiguous_sequence<Seq> &&
+            sized_sequence<Seq> &&
+            std::same_as<Value, value_t<Seq>> &&
+            // only allow memset on single byte types
+            sizeof(value_t<Seq>) == 1 &&
+            std::is_trivially_copyable_v<value_t<Seq>>;
+
+        if constexpr (can_memset) {
+            if (std::is_constant_evaluated()) {
+                impl(seq, value);
+            } else {
+                auto size = flux::usize(seq);
+                if(size == 0){
+                    return;
+                }
+                
+                FLUX_ASSERT(flux::data(seq) != nullptr);
+                
+                std::memset(flux::data(seq), value,
+                    flux::usize(seq) * sizeof(value_t<Seq>));
```

Comment:
We've already got the size, we may as well re-use it

```suggestion
                std::memset(flux::data(seq), value,
                    size * sizeof(value_t<Seq>));
```